Adding new features for XRM v8.2

clownwilleatme opened the pull request:
There is a breaking change with the Control interface which may not be desirable. However, it seems to be required as Microsoft has changed things quite a bit with the latest XRM release (e.g. the introduction of the QuickForm control which does not support many of the methods that most other controls do).

## Adding new features for XRM v8.2
1. **Xrm.Page.ui**
   * Adding quickForms collection
   * Adding Xrm.Page.ui.QuickForm (https://msdn.microsoft.com/en-us/library/mt736908.aspx)
2. **Xrm.Async & XrmPromise**
   * Modifying the Async definitions to allow methods to return specific success or error callback delegates.
   * This is required because of new methods which return promises using callback delegates with different parameters (e.g. Xrm.Utility.openQuickCreate and Xrm.Mobile.offline methods)
3. **Xrm.Utility**
   * Modifying the Xrm.Utility.openQuickCreate method and related asynchronous promise methods (https://msdn.microsoft.com/en-us/library/jj602956.aspx#BKMK_openQuickCreate)
4. **Xrm.Page.data**
   * Adding Xrm.Page.data.save method with saveOptions (https://msdn.microsoft.com/en-us/library/dn481607.aspx#Anchor_2)
5. **Controls**
   * Moved methods from the Control interface into the StandardControl interface. This will be a breaking change, but is required because the Xrm.Page.ui.getControl methods can now return quick-form controls which support a smaller sub-set of methods than other controls.
   * Created UiLabelElement interface to controls which have labels which can be viewed and updated.
   * Created UiCanGetVisibleElement interface for controls which can have their current visibility level viewed (all form controls)
   * Created UiStandardElement interface for standard controls. This includes controls such as StandardControl, Tab and Section, but does not apply to the new quick form controls as they cannot have their visibility set.
   * Updated Control.getControlType method to use new type for ControlType with pre-defined string values.
   * Moving methods from AutoLookupControlinto UiKeyPressable interface as they are supported for both string and number controls (ref: https://msdn.microsoft.com/en-us/library/gg334266.aspx#Anchor_8)
   * Creating StringControl interface, extending the AutoLookupControl class to avoid breaking existing code.
   * Creating NumberControl interface.
6. **Attributes**
   * Updating the base Attribute interface to add getValue and setValue methods so that the attribute value can be accessed without casting to a data specific Attribute which is useful when the attribute type is not known. As per MSDN, all attributes have the getValue and setValue methods, but the return type depends on the attribute type (https://msdn.microsoft.com/en-us/library/gg334409.aspx#Anchor_15)
   * Modifying NumberAttribute, StringAttribute, DateAttribute, OptionSetAttribute, and LookupAttribute to return the relevant control collections (e.g. StringAttribute.controls returns Collection.ItemCollection<StringControl>)
   * Updated Attribute.getAttributeType and getFormat method to use new types AttributeType and AttributeFormat with pre-defined string values.
   * Updated getFormat method in DateAttribute, NumberAttribute, OptionSetAttributeFormat and StringAttribute to only return relevant values based on types for each attribute type (e.g. StringAttributeType).
7. **Xrm.Mobile.offline**
   * Adding new Xrm.Mobile.Offline (https://msdn.microsoft.com/en-us/library/mt787123.aspx)
8. **Xrm.Panel**
   * Adding new Xrm.Panel methods (https://msdn.microsoft.com/en-us/library/mt790281.aspx)
9. **Other**
   * Adding additional comments
   * Formatting changes to keep things consistent

Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

daryllabar commented:
As much as I applaud @clownwilleatme for the documentation of what change, Github is showing every single line as being changed, and as such I'm not willing to review this change.

aozgaa commented:
Perhaps it is a CRLF issue? Please make the diff more readable.

clownwilleatme commented:
Sorry, should be fixed now, I should have checking properly first.

daryllabar commented on xrm/index.d.ts:
```diff
@@ -205,16 +214,194 @@ declare namespace Xrm
+        /**
+          * Provides methods to create and manage records in the mobile clients (for phones tablets).
+          */
+        Mobile: {
```
I'm guessing this should be an interface, not just a field.

clownwilleatme replied:
Sure, I'll make that change. Any thoughts on the changes to the Control interface?

daryllabar replied:
This should be an interface, not just a field.

daryllabar replied:
You mean how you've split them into smaller interfaces?

daryllabar commented on xrm/index.d.ts:
```diff
@@ -394,11 +605,27 @@ declare namespace Xrm
+          *                                       WARNING Notification will use the system warning icon.
+          *                                       INFO    Notification will use the system info icon.
+          */
+        setFormNotification( message: string, level: Page.ui.FormNotificationLevel | string, uniqueId: string ): boolean;
```
Thinking we should remove the union type string   ERROR, WARNING and INFO are the only options.

daryllabar replied:
Thinking we should remove the union type string since ERROR, WARNING and INFO are the only options.

daryllabar commented on xrm/index.d.ts:
```diff
@@ -1111,7 +1502,7 @@ declare namespace Xrm
              * @param   {string}    requirementLevel    The requirement level, as either "none", "required", or "recommended"
              */
-            setRequiredLevel(requirementLevel: RequirementLevel | string): void;
+            setRequiredLevel( requirementLevel: RequirementLevel | string ): void;
```
Again, thinking we should remove the string union type.

clownwilleatme replied:
Agreed, same for `setSubmitMode( submitMode: SubmitMode | string ): void;`

daryllabar replied:
Again, thinking we should remove the string union type.

daryllabar replied:
yep

daryllabar commented on xrm/index.d.ts:
```diff
@@ -1677,16 +2145,12 @@ declare namespace Xrm
-             */
-            getDisabled(): boolean;
-
+            getControlType(): ControlType | string;
```
Same here.  I know it is a breaking change, but it would be better to force the correct value imho

clownwilleatme replied:
This one needs to be a union type because it can return dynamic control names.

daryllabar replied:
Ah yah.

daryllabar commented on xrm/index.d.ts:
```diff
@@ -205,16 +214,194 @@ declare namespace Xrm
+              *
+              * @see {@link https://msdn.microsoft.com/en-us/library/mt787123.aspx} for details.
+              */
+            offline: {
```
Interface here to as well

clownwilleatme commented:
I have made the changes. However, removing the string union types has caused an error for one of the tests.

`let requirementLevelString = "none";
attribute.setRequiredLevel(requirementLevelString); // This fails
attribute.setRequiredLevel("none"); // This works`

Do you still think removing the string unions is worth it?

clownwilleatme commented:
@daryllabar I have made the changes requested, but there's still a question about having string unions for various methods. Having to define strings as consts to pass to methods like setRequiredMode does remove a bit of flexibility, IMO.

daryllabar commented:
You're not having to define them as constants, just cast them as that union type or any.  They will only affect someone that is using a variable to pass to the method, which I would guess would only be in a shared function, so it should be a pretty minimal breaking change.

But I'm definitely not going to require it.  If you don't agree with it, don't do it.

clownwilleatme commented:
Yeah, you're right - it's a pretty small change and declaring as the union type more or less negates the issue. Are you happy with everything else?

What's the next step for getting this PR approved?

aozgaa commented on xrm/index.d.ts:
```diff
@@ -394,11 +452,27 @@ declare namespace Xrm
+          * Displays a form level notification. Any number of notifications can be displayed and will remain until removed using clearFormNotification.
+          * The height of the notification area is limited so each new message will be added to the top.
+          *
+          * @param   {string}                            message     The text of the notification message.
```
You don't need the types in jsdoc comments if you are declaring them below.

clownwilleatme replied:
I was just following the existing format that has been used for most of the other methods. Happy to remove them though.

daryllabar commented:
Looks good to me.  I Thumbed up the review which I believe means someone will get around to merging the PR eventually.

aozgaa commented on xrm/index.d.ts:
```diff
@@ -474,6 +549,179 @@ declare namespace Xrm
+          * @return   {Async.XrmPromise}        Returns an asynchronous promise.
+          *
+          * @remarks  You cannot create intersect and activity party entities.
+          *           Only the following attribute types are supported in offline mode: BigInt
```
Can you start the list on a separate line so it doesn't need to be indented so much?

aozgaa commented on xrm/index.d.ts:
```diff
@@ -474,6 +549,179 @@ declare namespace Xrm
+          * @return   {Async.XrmPromise} Returns an asynchronous promise.
+          *
+          * @remarks  Only $select option can be specified within $expand.
+          *           Only the following attribute types are supported in offline mode: BigInt
```
ditto

aozgaa commented on xrm/index.d.ts:
```diff
@@ -474,6 +549,179 @@ declare namespace Xrm
+          * @remarks  A maximum of 5000 related records can be retrieved when using $expand.
+          *           Only $select option can be specified within $expand.
+          *           $skip query option is not supported.
+          *           Only the following attribute types are supported in offline mode: BigInt
```
Seeing as this remark is copied so much, I wonder if there might be some way to de-duplicate it.

daryllabar replied:
I'm not aware of any way to share comments....

clownwilleatme replied:
I tried to think of a way to do that but came up blank. Maybe putting them in a few lines instead of one per line will make it a bit easier to read?

aozgaa commented:
Can you add a `tslint.json` to make the formatting consistent?

clownwilleatme commented:
@aozgaa I have added a tslint file with rules that fit the existing formatting. I have updated the various files to correct minor formatting errors.

For the Xrm.Mobile.offline methods with the big lists of supported data types, I have reformatted them so that they take up much less space. Does it look good to you?

aozgaa commented:
Can you use a `tslint.json` that follows the general repo rules? ie:
```
{
    "extends": "../tslint.json",
    "rules": {
        // Special rules you *really* need
    }
}
```

aozgaa commented:
Actually let's just do it in a separate PR since this looks functionally correct.
